x-pack/plugins/fleet/server/services/secrets.ts

```diff
@@ -0,0 +1,296 @@
+  NewPackagePolicy,
+  PackagePolicyConfigRecordEntry,
+  RegistryStream,
+} from '../../common';
+
+import {
+  doesPackageHaveIntegrations,
+  getNormalizedDataStreams,
+  getNormalizedInputs,
+} from '../../common/services';
+
+import type {
+  PackageInfo,
+  PackagePolicy,
+  RegistryVarsEntry,
+  Secret,
+  VarSecretReference,
+  PolicySecretReference,
+} from '../types';
+
+import { FleetError } from '../errors';
+import { SECRETS_INDEX } from '../constants';
+
+import { auditLoggingService } from './audit_logging';
+
+import { appContextService } from './app_context';
+
+interface SecretPath {
+  path: string;
+  value: PackagePolicyConfigRecordEntry;
+}
+
+// This will be removed once the secrets index PR is merged into elasticsearch
+function getSecretsIndex() {
+  const testIndex = appContextService.getConfig()?.developer?.testSecretsIndex;
+  if (testIndex) {
+    return testIndex;
+  }
+  return SECRETS_INDEX;
+}
+
+export async function createSecrets(opts: {
+  esClient: ElasticsearchClient;
+  values: string[];
+}): Promise<Secret[]> {
+  const { esClient, values } = opts;
+  const logger = appContextService.getLogger();
+  const body = values.flatMap((value) => [
+    {
+      create: { _index: getSecretsIndex() },
+    },
+    { value },
+  ]);
+  let res: BulkResponse;
+  try {
+    res = await esClient.bulk({
+      body,
+    });
+
+    const [errorItems, successItems] = partition(res.items, (a) => a.create?.error);
+
+    successItems.forEach((item) => {
+      auditLoggingService.writeCustomAuditLog({
+        message: `secret created: ${item.create!._id}`,
+        event: {
+          action: 'secret_create',
+          category: ['database'],
+          type: ['access'],
+          outcome: 'success',
+        },
+      });
+    });
+
+    if (errorItems.length) {
+      throw new Error(JSON.stringify(errorItems));
+    }
+
+    return res.items.map((item, i) => ({
+      id: item.create!._id as string,
+      value: values[i],
+    }));
+  } catch (e) {
+    const msg = `Error creating secrets in ${getSecretsIndex()} index: ${e}`;
+    logger.error(msg);
+    throw new FleetError(msg);
+  }
+}
+
+export async function deleteSecret(opts: {
+  esClient: ElasticsearchClient;
+  id: string;
+}): Promise<DeleteResponse['result']> {
+  const { esClient, id } = opts;
+  let res: DeleteResponse;
+  try {
+    res = await esClient.delete({
+      index: getSecretsIndex(),
+      id,
+    });
+
+    auditLoggingService.writeCustomAuditLog({
+      message: `secret deleted: ${id}`,
+      event: {
+        action: 'secret_delete',
+        category: ['database'],
+        type: ['access'],
+        outcome: 'success',
+      },
+    });
+  } catch (e) {
+    const logger = appContextService.getLogger();
+    const msg = `Error deleting secret '${id}' from ${getSecretsIndex()} index: ${e}`;
+    logger.error(msg);
+    throw new FleetError(msg);
+  }
+
+  return res.result;
+}
+
+export async function extractAndWriteSecrets(opts: {
```

Comment:
this could be covered with unit test

Reply:
This is covered by the integration tests, I think the unit tests would need quite a lot of mocking